apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -101,6 +101,12 @@ protected Deque<ElasticContext<?>> initialValue() {
         }
     };
 
+    private final ThreadLocal<StackOverflowCounter> activeStackOverflowCounter = new ThreadLocal<StackOverflowCounter>() {
```

Comment:
I think we could make this a bit cleaner by creating a dedicated `ActiveStack` class that has both an `ArrayDeque<ElasticContext<?>>` and an overflow counter. This abstracts the overflow handling so that the activation/deactivation methods don't need to deal with that.